pkg/reconciler/managed/reconciler.go

```diff
@@ -693,17 +726,25 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	// not realize that the controller is still trying to reconcile
 	// (and modify or delete) the resource since they forgot to enable the
 	// feature flag.
-	if !r.managementPoliciesEnabled && (managed.GetManagementPolicy() == xpv1.ManagementObserveOnly || managed.GetManagementPolicy() == xpv1.ManagementOrphanOnDelete) {
-		log.Debug(errManagementPolicy, "policy", managed.GetManagementPolicy())
-		record.Event(managed, event.Warning(reasonManagementPolicyNotEnabled, errors.New(errManagementPolicy)))
-		managed.SetConditions(xpv1.ReconcileError(errors.New(errManagementPolicy)))
+	if isNonDefaultManagementPolicy(managementPoliciesEnabled, managed.GetManagementPolicy()) {
```

Comment:
Would these sanity checks (e.g. using management policies without the feature flag, using invalid management policies) be suited to a webhook? Or perhaps an Initializer? Basically wondering if we can push them out of this (already very complex) reconcile loop.

Reply:
If it was only up to the managed resources, then yeah, we could. But as its also dependent on the flag in the provider, which can change over time, I am not sure if we can catch for example a case:
1. Management policies enabled
2. Add resource with non-default management policy - passes webhook/whatever
3. Management policies disabled - we have no way now to sanity check the above resource

Here in the reconciler we have the only place where both the flag and the resource come together live.

Probably something like OPA would be able to validate the provider flag change and warn/disallow it if resources are using custom managment policies, but that seems like an overkill. The code will get simpler when we reach GA, although I realize that is not soon.

Reply:
Yeah, this is probably wishful thinking on my part.